File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.java


class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // 1. Count frequencies using a HashMap
        Map<Integer, Integer> countMap = new HashMap<>();
        for (int num : nums) {
            countMap.put(num, countMap.getOrDefault(num, 0) + 1);
        }

        // 2. Create a list of your custom 'Element' objects
        List<Element> list = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : countMap.entrySet()) {
            list.add(new Element(entry.getKey(), entry.getValue()));
        }

        // 3. Sort the list (Collections.sort uses your compareTo logic)
        // We want descending order (highest frequency first)
        Collections.sort(list);

        // 4. Extract the top K 'actual' values
        int[] ans = new int[k];
        for (int i = 0; i < k; i++) {
            ans[i] = list.get(i).actual;
        }
        return ans;
    }

    // Fixed: Must implement Comparable<Element> for sorting to work
    class Element implements Comparable<Element> {
        int actual;
        int freq;

        public Element(int act, int f) {
            actual = act;
            freq = f;
        }

        @Override
        public int compareTo(Element l) {
            // Sort by frequency in descending order
            return l.freq - this.freq;
        }
    }
}

File: Data Structures & Algorithms/house-robber/submission-0.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums)<2:
            if len(nums) == 0:
                return 0
            elif len(nums) == 1:
                return nums[0]
            else:
                return max(nums[0], num[1])
        n = len(nums)
        answer = [0] * n
        answer[n-1] = nums[n-1]
        answer[n-2] = max(nums[n-1], nums[n-2])
        for i in range (len(nums)-3, -1, -1):
            if (nums[i] + answer[i+2]) > answer[i+1]:
                answer[i] = nums[i] + answer[i+2]
            else:
                answer[i] = answer[i+1]

        return answer[0]
            
